### ISSUE TITLE
How do I get the hostname?

### ISSUE THREAD

**Issue by roddc** [haf-decent/arpping#15]
Hi, I would like to get the hostname of the found devices by using `discover` method, thanks for your effort.

**Comment by haf-decent** [comment]
Hi @roddc, I just published a branch I created called hostnames. Let me know if you can clone that branch and run a discover scan. It just appends the hostname specified by the arp output to each found host as the `name` property. It seems like a useful addition, so I'll fold in these changes to a new npm release once you've had a chance to review.

**Comment by roddc** [comment]
Hi, I test your branch, but I got 'Address' as name for each found host. Is this result correct?

**Comment by haf-decent** [comment]
Oh, sorry, I'm only able to test on macOS right now, so the string parsing is probably not correct for the arp output on your OS. Could you just open up a command line terminal, arp an ip on your network, and send me the printed output so I can update the parser?

**Comment by roddc** [comment]
```
Address                  HWtype            Flags Mask            Iface
10.53.9.21               ether                 C                     enp3s0
```

It seems arp couldn't resolve the hostname. Do I need to configure something? I test it on Ubuntu 18.04

**Comment by roddc** [comment]
Hi, I also found an error on Ubuntu that uses Chinese language, the error says in method `arp`, the variable 'mac' is undefined. I think the problem is `stdout.includes('no entry')`, the 'no entry' will be translated into other languages

**Comment by haf-decent** [comment]
It looks like you can do `nslookup <ip>` to find a hostname on most os's. Can you test and make sure that works on your end? I think it's a standard library, but I might want to add a shell script or something to make sure all the required command line libraries are installed anyway.

And then for multi-language support, you're entirely right. I should probably be using regex parsers anyway to extract mac addresses rather than relying on language specific messaging. I'll need to spend a bit of time to get that working properly.

**Comment by haf-decent** [comment]
Ok, just pushed some changes to the `hostnames` branch. I added the `nslookup <ip>` command + regex parser (which may need work) to arp, as well as a regex parser for the mac address. The ping parser still looks for the English string "100% packet loss" to determine if a host is unresponsive/does not exist, so I'll probably need to find a different method for that.

Please pull those changes when you can and let me know if that has addressed any of the above issues.

**Comment by roddc** [comment]
I test the branch, it's working. However, I have another issue on Windows, it seems `arp` on Windows need a flag `-a` otherwise no result will be displayed. Also, the `getFullRange` method on Windows will return 10K+ items, which casues the application freeze. Is there a way to set the subnet mask?

**Comment by roddc** [comment]
> Ok, just pushed some changes to the `hostnames` branch. I added the `nslookup <ip>` command + regex parser (which may need work) to arp, as well as a regex parser for the mac address. The ping parser still looks for the English string "100% packet loss" to determine if a host is unresponsive/does not exist, so I'll probably need to find a different method for that.
> 
> Please pull those changes when you can and let me know if that has addressed any of the above issues.

I found a minor bug in `nslookup` callback, `const [ name = null ] = stdout.match(/ = .*/);` should be `const [ name = null ] = stdout.match(/ = .*/) || [];`

**Comment by haf-decent** [comment]
Oh, right, thank you, I will update the nslookup command. Does `arp -a` return a mac address on windows? The example output you sent me doesn't seem to include one.

Also, this is what it looks like when I run `nslookup 192.168.0.181`:
`Server:		192.168.0.1`
`Address:	192.168.0.1#53`

`181.0.168.192.in-addr.arpa	name = hostname.`

I was hoping the ` = ` part would be consistent across os's regardless of language, but maybe that was a bit too optimistic. Could you let me know exactly what your command line output looks like for that so I can make sure the parser is consistent?

**Comment by roddc** [comment]
On Ubuntu 18.04(English):
`arp -a arp -a 10.53.9.158`
`? (10.53.9.158) at 9c:4b:0d:6c:e1:49 [ether] on enp3s0`

`nslookup 10.53.9.158`
`** server can't find 158.9.53.10.in-addr.arpa: NXDOMAIN`

On Windows 10(Chinese):
`arp -a 10.53.8.208`
```
接口： 10.53.8.169--0x14
 Internet 地址      物理地址                 类型
 10.53.8.208       16-5b-fc-6b-61-f5     动态
```

`nslookup 10.53.9.158`
```
服务器: Unknown
 Address: 10.53.5.5
*** Unknown 找不到 10.53.9.158: Non-existent domain
```

Above examples use other devices'ip, not the host ip
and it seems `arp` cannot find the host ip

**Comment by haf-decent** [comment]
Ok, the Windows 10 Chinese example seems consistent with what I expected, although I'm still just hoping the ` = <hostname>` format is the same for `nslookup` when it does resolve a hostname. I just pushed some updates, including adding the `-a` flag for windows and allowing for custom netmask definitions for `_getFullRange` that you can then use in the discover scan:

```javascript
const arpping = new Arrping();

const customRange = arpping._getFullRange(24);
arpping.discover(customRange);

arpping.discover(
    arpping._getFullRange('255.255.255.0')
);
```

I hope that fixes the 10k+ address freeze issue you were having, at least as a bandaid solution. Let me know if we're getting closer.

**Comment by roddc** [comment]
> Ok, the Windows 10 Chinese example seems consistent with what I expected, although I'm still just hoping the ` = <hostname>` format is the same for `nslookup` when it does resolve a hostname. I just pushed some updates, including adding the `-a` flag for windows and allowing for custom netmask definitions for `_getFullRange` that you can then use in the discover scan:
> 
> ```js
> const arpping = new Arrping();
> 
> const customRange = arpping._getFullRange(24);
> arpping.discover(customRange);
> 
> arpping.discover(
>     arpping._getFullRange('255.255.255.0')
> );
> ```
> 
> I hope that fixes the 10k+ address freeze issue you were having, at least as a bandaid solution. Let me know if we're getting closer.

Thanks a lot, that's what I looking for

**Comment by haf-decent** [comment]
Ok, well I released a new version today that includes the `nslookup` call. All hosts now have the `name` property, which resolves to a string or null.

So this should hopefully be resolved with v4.0.0. Gonna mark as closed.
